File: Source/ObjC/NSKTopImageLayout.m
//
//  NSKTopImageLayout.m
//  NSKButtonObjC
//
//  Created by NSSimpleApps on 21.08.16.
//  Copyright © 2016 NSSimpleApps. All rights reserved.
//

#import "NSKTopImageLayout.h"

@implementation NSKTopImageLayout

+ (CGRect)titleRectForContentRect:(CGRect)contentRect
                 defaultTitleRect:(CGRect)defaultTitleRect
                        imageRect:(CGRect)imageRect
                  titleEdgeInsets:(UIEdgeInsets)titleEdgeInsets {
    
    if (CGRectIsEmpty(defaultTitleRect)) {
        
        return [super titleRectForContentRect:contentRect
                             defaultTitleRect:defaultTitleRect
                                    imageRect:imageRect
                              titleEdgeInsets:titleEdgeInsets];
        
    } else {
        
        CGPoint p =
        CGPointMake(CGRectGetMidX(contentRect) - defaultTitleRect.size.width/2 + titleEdgeInsets.left - titleEdgeInsets.right,
                    CGRectGetMaxY(contentRect) - defaultTitleRect.size.height + titleEdgeInsets.top - titleEdgeInsets.bottom);
        
        return (CGRect){.origin = p, .size = defaultTitleRect.size};
    }
}

+ (CGRect)imageRectForContentRect:(CGRect)contentRect
                 defaultImageRect:(CGRect)defaultImageRect
                        titleRect:(CGRect)titleRect
                  imageEdgeInsets:(UIEdgeInsets)imageEdgeInsets {
    
    if (CGRectIsEmpty(defaultImageRect)) {
        
        return [super imageRectForContentRect:contentRect
                             defaultImageRect:defaultImageRect
                                    titleRect:titleRect
                              imageEdgeInsets:imageEdgeInsets];
        
    } else {
        
        return CGRectMake(CGRectGetMidX(contentRect) - defaultImageRect.size.width/2 + imageEdgeInsets.left - imageEdgeInsets.right,
                          CGRectGetMinY(contentRect) + imageEdgeInsets.top - imageEdgeInsets.bottom,
                          defaultImageRect.size.width,
                          defaultImageRect.size.height);
    }
}

@end

File: Source/Swift/NSKTopImageLayout.swift
//
//  NSKTopImageLayout.swift
//  NSKButtonSwift
//
//  Created by NSSimpleApps on 24.08.16.
//  Copyright © 2016 NSSimpleApps. All rights reserved.
//

import UIKit


internal class NSKTopImageLayout: NSKVerticalImageLayout {
    
    override static func titleRect(forContentRect contentRect: CGRect,
                                  defaultTitleRect: CGRect,
                                  imageRect: CGRect,
                                  titleEdgeInsets: UIEdgeInsets) -> CGRect {
        
        let p =
            CGPoint(x: contentRect.midX - defaultTitleRect.width/2 + titleEdgeInsets.left - titleEdgeInsets.right,
                    y: contentRect.maxY - defaultTitleRect.height + titleEdgeInsets.top - titleEdgeInsets.bottom)
        
        return CGRect(origin: p, size: defaultTitleRect.size)
    }
    
    override static func imageRect(forContentRect contentRect: CGRect,
                                  defaultImageRect: CGRect,
                                  titleRect: CGRect,
                                  imageEdgeInsets: UIEdgeInsets) -> CGRect {
        
        return CGRect(x: contentRect.midX - defaultImageRect.width/2 + imageEdgeInsets.left - imageEdgeInsets.right,
                      y: contentRect.minY + imageEdgeInsets.top - imageEdgeInsets.bottom,
                      width: defaultImageRect.width,
                      height: defaultImageRect.height)
    }
}
